### PR TITLE
Refactor onEvents and IsSecure for TextField

### PR DESCRIPTION
**Pull request by jcaracciolo**
- Now TextField has a IsSecure property that would display it as if it where a password textField
- Now TextField has a shouldReturn property that would dismiss the editing when the enter key is pressed
- Now TextField receives a closure in order to be able to pass the text inside the textfield for any TextFieldEvent

### CODE REVIEW

File: Portal/View/Components/TextField.swift [hunk 1]
```diff
@@ -12,60 +12,57 @@ public struct TextFieldProperties<MessageType> {
...
-            onEvents: self.onEvents.map(transform)
+            isSecureTextEntry: isSecureTextEntry,
+            onEvents: { return self.onEvents($0).map(transform) }
         )
     }
     
 }
 
-public struct TextFieldEvents<MessageType> {
+public enum TextFieldEvents {
```
Comment (guidomb) [inline Portal/View/Components/TextField.swift]: Why did you change this to an enum? What happens if a user wants to dispatch a message `onEditingBegin` and `onEditingEnd`? Would it make sense?

File: Portal/View/Components/TextField.swift [hunk 2]
```diff
@@ -12,60 +12,57 @@ public struct TextFieldProperties<MessageType> {
...
+    case onEditingChanged(text: String)
+    case onEditingEnd(text: String)
     
-    public func map<NewMessageType>(_ transform: (MessageType) -> NewMessageType) -> TextFieldEvents<NewMessageType> {
-        return TextFieldEvents<NewMessageType>(
-            onEditingBegin: onEditingBegin.map(transform),
-            onEditingChanged: onEditingChanged.map(transform),
-            onEditingEnd: onEditingEnd.map(transform)
-        )
+    static func fromUI(_ event: UIControlEvents, text: String) -> TextFieldEvents? {
```
Comment (guidomb) [inline Portal/View/Components/TextField.swift]: This method should not be here. It should be inside the renderer as a helper function because it depends on UIKit.

File: Portal/View/UIKit/Renderers/TextFieldRenderer.swift [hunk 3]
```diff
@@ -29,30 +30,32 @@ internal struct TextFieldRenderer<MessageType, RouteType: Route>: UIKitRenderer
...
-                message: $0,
-                for: .editingChanged,
-                with: mailbox)
-            }
-            properties.onEvents.onEditingEnd |> { _ = textField.dispatch(
-                message: $0,
-                for: .editingDidEnd,
-                with: mailbox)
-            }
+        let dispatcher: (Mailbox<ActionType>) -> Void = { mailbox in
```
Comment (guidomb) [inline Portal/View/UIKit/Renderers/TextFieldRenderer.swift]: Why are you using a variable to store the closure instead of passing this closure directly to `bindMessageDispatcher`?

File: Portal/View/UIKit/Renderers/TextFieldRenderer.swift [hunk 4]
```diff
@@ -29,30 +30,32 @@ internal struct TextFieldRenderer<MessageType, RouteType: Route>: UIKitRenderer
         textField.removeTarget(.none, action: .none, for: .editingDidBegin)
         textField.removeTarget(.none, action: .none, for: .editingChanged)
         textField.removeTarget(.none, action: .none, for: .editingDidEnd)
+        if properties.shouldReturn {
+            textField.delegate = TextFieldDelegate.shared
```
Comment (guidomb) [inline Portal/View/UIKit/Renderers/TextFieldRenderer.swift]: Delegate have weak references so what about making `UITextField` its own delegate by confirming to the protocol.

**Review by guidomb — CHANGES_REQUESTED** [review]

File: Portal/View/Components/TextField.swift (hunk 1 above)
Comment (jcaracciolo) [reply Portal/View/Components/TextField.swift]: Because there is a need to be sent the string of the current textField. As you cant force a messageType to containe a String, the best way is to ask for a function that receives these enums, with the respective string, and return a MessageType

File: Portal/View/UIKit/Renderers/TextFieldRenderer.swift (hunk 4 above)
Comment (jcaracciolo) [reply Portal/View/UIKit/Renderers/TextFieldRenderer.swift]: You mean with an extension of UITextField?

Comment (guidomb) [reply Portal/View/UIKit/Renderers/TextFieldRenderer.swift]: Yes

File: Portal/View/Components/TextField.swift (hunk 1 above)
Comment (guidomb) [reply Portal/View/Components/TextField.swift]: I would change `TextFieldEvents` to

```swift
struct TextFieldEvents<MessageType> {

    let onEditingBegin: ((String) -> MessageType)?
    let onEditingChanged: ((String) -> MessageType)?
    let onEditingEnd: ((String) -> MessageType)?

    fileprivate init(
        onEditingBegin: ((String) -> MessageType)? = .none,
        onEditingChanged: ((String) -> MessageType)?,
        onEditingEnd: ((String) -> MessageType)? = .none
    ) {
        self.onEditingBegin = onEditingBegin
        self.onEditingChanged = onEditingChanged
        self.onEditingEnd = onEditingEnd
    }

}
```

and then I would add a builder function in order to do something like

```swift
textFieldEvents() {
     $0.onEditingBegin = { .myCustomMessage($0) }
}
```

**Review by guidomb — APPROVED** [review]
